articles/active-directory/develop/msal-net-client-assertions.md

```diff
@@ -116,30 +123,30 @@ string Encode(byte[] arg)
     return s;
 }
 
-string GetSignedClientAssertion()
+static string GetSignedClientAssertion(string tenantId, string clientId)
 {
-    //Signing with SHA-256
-    string rsaSha256Signature = "http://www.w3.org/2001/04/xmldsig-more#rsa-sha256";
-     X509Certificate2 certificate = new X509Certificate2("Certificate.pfx", "Password", X509KeyStorageFlags.EphemeralKeySet);
+    // It's not recommended to hardcode a password or a file location. Get the X509Certificate2 from somewhere save.
+    X509Certificate2 certificate = new X509Certificate2("Certificate.pfx", "Password", X509KeyStorageFlags.EphemeralKeySet);
 
-    //Create RSACryptoServiceProvider
-    var x509Key = new X509AsymmetricSecurityKey(certificate);
-    var privateKeyXmlParams = certificate.PrivateKey.ToXmlString(true);
-    var rsa = new RSACryptoServiceProvider();
-    rsa.FromXmlString(privateKeyXmlParams);
+    // Get the RSA with the private key, used for signing.
+    var rsa = certificate.GetRSAPrivateKey();
 
     //alg represents the desired signing algorithm, which is SHA-256 in this case
     //kid represents the certificate thumbprint
     var header = new Dictionary<string, string>()
-         {
-              { "alg", "RS256"},
-              { "kid", Encode(certificate.GetCertHash()) }
-         };
+    {
```

Comment:
I'm getting some errors here in some configurations (i run the test on a pre-prod env as well as production). I think the header should instead be: 
```csharp
var header = new Dictionary<string, string>()
            {
              { "alg", "RS256"},
              { "kid",  certificate.Thumbprint},
              { "typ", "JWT"},
              { "x5t", Encode(certificate.GetCertHash())},
            };
```

Reply:
According to the [msal internal code](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/ef3b2f61d9f0d441a1d2232ca2062a8d60e13437/src/client/Microsoft.Identity.Client/Internal/JsonWebToken.cs#L212-L235) the `x5t` parameter is mandatory for **ADFS** and the `kid` is optional.

According to the [Azure AD docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials) it also only expects a `x5t`, so that must be the way to go.

But apart from that I did test the situation with only the above header (`alg` and `kid` set to base64 encoded cert hash) to be working against Azure AD.

Reply:
Yes, I got it working as well with only `alg` and `kid` on AAD but that failed on ADFS. So it's safer to include both `kid` and `x5t`. 

The IDP needs to find the public key of the cert used to sign this assertion in order to validate the signature. It is possible that you have N certs uploaded to the IDP. So it's going to use the `x5t` to find the cert. 

I'm sure AAD has built some smarts into this and can handle the `kid` only case, but the recommendation is to send `x5t`. Probably faster as well.